From 2b12d3e52eef333fa79c48f0cbb30849eaec9dd9 Mon Sep 17 00:00:00 2001
From: Adam Welle <arwelle@cert.org>
Date: Tue, 3 Nov 2015 15:50:29 -0500
Subject: [PATCH 2/3] mac80211_hwsim: adds ATTR_FREQ to netlink messages and
 checks on receive

This updates mac80211_hwsim.c and mac80211_hwsim.h to include the
HWSIM_ATTR_FREQ value.
---
 drivers/net/wireless/mac80211_hwsim.c | 14 ++++++++++++++
 drivers/net/wireless/mac80211_hwsim.h |  1 +
 2 files changed, 15 insertions(+)

diff --git a/drivers/net/wireless/mac80211_hwsim.c b/drivers/net/wireless/mac80211_hwsim.c
index 0e9cc78..362e38e 100644
--- a/drivers/net/wireless/mac80211_hwsim.c
+++ b/drivers/net/wireless/mac80211_hwsim.c
@@ -411,6 +411,7 @@ static struct nla_policy hwsim_genl_policy[HWSIM_ATTR_MAX + 1] = {
 				 .len = IEEE80211_TX_MAX_RATES*sizeof(
 					struct hwsim_tx_rate)},
 	[HWSIM_ATTR_COOKIE] = { .type = NLA_U64 },
+        [HWSIM_ATTR_FREQ] = { .type = NLA_U32 },
 };
 
 static netdev_tx_t hwsim_mon_xmit(struct sk_buff *skb,
@@ -661,6 +662,9 @@ static void mac80211_hwsim_tx_frame_nl(struct ieee80211_hw *hw,
 	if (nla_put_u32(skb, HWSIM_ATTR_FLAGS, hwsim_flags))
 		goto nla_put_failure;
 
+	if (nla_put_u32(skb, HWSIM_ATTR_FREQ, data->channel->center_freq))
+		goto nla_put_failure;
+
 	/* We get the tx control (rate and retries) info*/
 
 	for (i = 0; i < IEEE80211_TX_MAX_RATES; i++) {
@@ -2050,6 +2054,16 @@ static int hwsim_cloned_frame_received_nl(struct sk_buff *skb_2,
 
 	/*A frame is received from user space*/
 	memset(&rx_status, 0, sizeof(rx_status));
+
+	/* Check ATTR_FREQ if it exists, and throw away off-channel packets */
+	if (info->attrs[HWSIM_ATTR_FREQ]) {
+		rx_status.freq = nla_get_u32(info->attrs[HWSIM_ATTR_FREQ]);
+		if (rx_status.freq != data2->channel->center_freq)
+			goto out;
+	} else {
+		rx_status.freq = data2->channel->center_freq;
+	}
+
 	rx_status.freq = data2->channel->center_freq;
 	rx_status.band = data2->channel->band;
 	rx_status.rate_idx = nla_get_u32(info->attrs[HWSIM_ATTR_RX_RATE]);
diff --git a/drivers/net/wireless/mac80211_hwsim.h b/drivers/net/wireless/mac80211_hwsim.h
index afaad5a..95ac959 100644
--- a/drivers/net/wireless/mac80211_hwsim.h
+++ b/drivers/net/wireless/mac80211_hwsim.h
@@ -108,6 +108,7 @@ enum {
 	HWSIM_ATTR_SIGNAL,
 	HWSIM_ATTR_TX_INFO,
 	HWSIM_ATTR_COOKIE,
+	HWSIM_ATTR_FREQ,
 	__HWSIM_ATTR_MAX,
 };
 #define HWSIM_ATTR_MAX (__HWSIM_ATTR_MAX - 1)
-- 
2.1.4

